File: scripts/benchmark_settlementcheck.mjs
#!/usr/bin/env node
import { performance } from "node:perf_hooks";
import { stat } from "node:fs/promises";
import { SettlementCheck } from "../settlementcheck/js/settlementcheck.mjs";

const count = Number(process.argv[2] || 100000);
const started = performance.now();
const checker = await SettlementCheck.fromFile();
const loadMs = performance.now() - started;
let state = 2025;
const random = () => {
  state = (1664525 * state + 1013904223) >>> 0;
  return state / 2 ** 32;
};
const lons = Array.from({ length: count }, () => random() * 360 - 180);
const lats = Array.from({ length: count }, () => random() * 180 - 90);
for (let index = 0; index < Math.min(count,1000); index++) checker.classCode(lons[index], lats[index]);
const lookupStarted = performance.now();
for (let index = 0; index < count; index++) checker.classCode(lons[index], lats[index]);
const lookupMs = performance.now() - lookupStarted;
const artifact = new URL("../settlementcheck/data/degurba_R2025A_E2025_L12.tfdg", import.meta.url);
console.log(`artifact_bytes=${(await stat(artifact)).size}`);
console.log(`load_seconds=${(loadMs / 1000).toFixed(6)}`);
console.log(`scalar_queries_per_second=${Math.round(count / (lookupMs / 1000))}`);
console.log(`core_peak_rss_mib=${(process.resourceUsage().maxRSS/1024).toFixed(2)}`);
const detailStarted = performance.now();
await checker.checkAsync(-0.1276,51.5072);
console.log(`first_detail_face_seconds=${((performance.now()-detailStarted)/1000).toFixed(6)}`);
const warm = performance.now();
for(let i=0;i<10000;i++)checker.check(-0.1276,51.5072);
console.log(`warm_detail_queries_per_second=${Math.round(1e7/(performance.now()-warm))}`);
console.log(`with_detail_peak_rss_mib=${(process.resourceUsage().maxRSS/1024).toFixed(2)}`);

File: scripts/benchmark_settlementcheck.py
#!/usr/bin/env python
"""Measure settlementcheck load, scalar lookup and batch-convenience rates."""
from __future__ import annotations

import argparse
import random
import resource
import statistics
import sys
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / "settlementcheck" / "python"))
from settlementcheck import SettlementCheck  # noqa: E402


def timed(function, repeats=5):
    values = []
    for _ in range(repeats):
        started = time.perf_counter()
        function()
        values.append(time.perf_counter() - started)
    return statistics.median(values)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--data", type=Path,
                        default=ROOT / "settlementcheck/data/degurba_R2025A_E2025_L12.tfdg")
    parser.add_argument("-n", type=int, default=100_000)
    args = parser.parse_args()
    started = time.perf_counter()
    checker = SettlementCheck(args.data)
    load = time.perf_counter() - started
    core_load_rss = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1024
    rng = random.Random(2025)
    lons = [rng.uniform(-180, 180) for _ in range(args.n)]
    lats = [rng.uniform(-90, 90) for _ in range(args.n)]
    scalar = timed(lambda: [checker.class_code(lon, lat)
                            for lon, lat in zip(lons, lats)])
    batch = timed(lambda: checker.settlement_batch(lons, lats))
    print(f"artifact_bytes={args.data.stat().st_size}")
    print(f"runs={checker.stats['runs']} mixed_cells={checker.stats['mixed_cells']}")
    print(f"load_seconds={load:.6f}")
    print(f"core_load_peak_rss_mib={core_load_rss:.2f}")
    print(f"scalar_queries_per_second={args.n / scalar:.0f}")
    print(f"batch_convenience_points_per_second={args.n / batch:.0f}")
    print(f"core_peak_rss_mib={resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1024:.2f}")
    started = time.perf_counter()
    checker.check(-0.1276, 51.5072)
    print(f"first_detail_face_seconds={time.perf_counter() - started:.6f}")
    detail = timed(lambda: [checker.check(-0.1276, 51.5072) for _ in range(1000)])
    print(f"warm_detail_queries_per_second={1000 / detail:.0f}")
    print(f"with_detail_peak_rss_mib={resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1024:.2f}")


if __name__ == "__main__":
    main()
